tests/unit/test_storage.py

```diff
@@ -138,3 +141,60 @@ def test_reduced_chunk_fully_masked_data_vmax():
                          method=np.mean)
     assert rc[0].size == 0
     assert rc[1] is None
+
+
+def test_like_s3_reduce_chunk():
+    """Test reduce chunk entirely."""
+    rfile = "tests/test_data/cesm2_native.nc"
+    offset = 2
+    size = 128
+
+    # no compression
+    rc = st.reduce_chunk(rfile, offset, size,
+                         compression=None, filters=None,
+                         missing=[None, None, None, None],
+                         dtype=np.dtype("int32"), shape=(32, ),
+                         order="C", chunk_selection=slice(0, 2, 1),
+                         method=np.min)
+
+    assert rc[0] == 134351386
+    assert rc[1] == 2
+
+
+def test_s3_reduce_chunk():
```

Comment:
This is a unit test, I don't think it should require a running S3 server. That would be more appropriate for an integration test. Both are useful.

For a unit test, I'd mock out the requests library.

Reply:
Indeed, but in the meantime it has become a full on integration test since I am now uploading the file to the bucket and doing the whole s3 thing, man - check this out https://github.com/valeriupredoi/PyActiveStorage/actions/runs/5212808117/jobs/9406915387

Reply:
so yeah, I'll have to move it to integration testing. Question for you: the test is now reading the file from the S3 bucket on Minio, is there any difference between such a bucket and a (real) AWS one? Bceause as it appears, there are no issues with reading netCDF files from an S3 bucket :grin:

Reply:
Like I've said before, I don't think we need specific S3 integration tests. Just a set of tests that work with whatever storage backend has been configured.

Unit tests for the S3 integration are the right place to call directly into s3.reduce_chunk, and I've proposed some basic ones here: https://github.com/valeriupredoi/PyActiveStorage/pull/88/files

Reply:
Ow I think we need S3 integration tests - why wouldn't we? At this stage, I am just trying to navigate my way to see what we can do with S3 and netCDF, after that we can remove them

Reply:
I put some background on different types of testing here: https://github.com/valeriupredoi/PyActiveStorage/pull/88/#discussion_r1223440252.

This part is a bit more my personal opinion though:

In my view, we shouldn't need different integration tests for different storage backends. We should have a single set of integration tests that test PyActiveStorage against arbitrary storage backends. This is because both the interface and capabilities should be almost identical in each case. There are a few places where we'd need special cases for the specific storage backend:

* write test data to storage before performing the test
* pass an appropriate path/URL etc to `Active` to tell it how to retrieve the test data

Of course if testing against a real storage system then that needs to be available and some config provided to the test harness to tell it how to access it.

Otherwise though, the tests should be identical. This is how I set up the tests in test_bigger_data.py.

Perhaps the crucial part to make this click in practice is that we'd need multiple CI jobs.

1. unit tests (run `pytest tests/unit`)
2. integration tests with local storage (run `pytest tests/*.py`)
3. integration tests with S3 storage (setup minio & S3 active storage, run `pytest tests/*.py`)

Hopefully that helps explain what's in my head!

Now for holidays :sun_with_face:

Reply:
absolutely, that makes total sense - in a wee while, mate! Right now priority number one is figuring out if we can actually read the file at all from this or that storage, and if not - haul arse and see what we can do about it :grin: Have good holidays :beach_umbrella: